# Homework 3  - Tryout pandas and numpy

Copy this notebook. Rename it as: YOURNAME-Homework-panda-numpy-Sept-2017

with your name replacing yourname and  replaced with the date you submit or to the last part.

Do the homeworks problems below.

Upload your completed jupyter notebook to your github site and send me the url via the elearning site as your homework submission. Please note that you need to do your own 3.4, and 3.5 if you share data and analysis from others then you cannot get more than a 3.  

## Homework 3.1

### 3.1.a
Load the data from: http://opendata.dc.gov/datasets that I have include in this github   
into a dataframe.  ( The file has been is available in directory ./data/ccp_current_csv.csv )

In [31]:
import pandas as pd
data_frame=pd.read_csv("C:\Users\AsusGo\Desktop\ccp_current_csv.csv")
data_frame.head(5)


STREET QUADRANT  FROMINTERSECTION              TOINTERSECTION  \
0           13th St       NW        Kennedy St               Longfellow St   
1              S St       NW  North Capitol St                      3rd St   
2  Rhode Island Ave       NE  Brentwood Rd (W)            Brentwood Rd (M)   
3           Mill Rd       NW           27th St                  Cul De Sac   
4  Rhode Island Ave       NE  Brentwood Rd (M)  14th St  Evarts St  Montan   

   NOOFBLOCKS                 WORKDESCRIPTION  YEARBUDGETED  PERCENTCOMPLETED  \
0           1  Pavement Resurface/restoration          2004               100   
1           3                   Regular Cover          1999               100   
2           1  Pavement Resurface/restoration          2004               100   
3           1  Pavement Resurface/restoration          2004               100   
4           2  Pavement Resurface/restoration          2005               100   

      STATUS  MILES      ...             YCOORD   LATITUDE  LONGITUDE  \
0  Completed   0.07      ...       143205.11981  38.956747 -77.029694   
1  Completed   0.33      ...       138468.90256  38.914085 -77.012411   
2  Completed   0.09      ...       139574.10659  38.924041 -76.987221   
3  Completed   0.07      ...       138132.90198  38.911046 -77.054077   
4  Completed   0.16      ...       139657.25318  38.924790 -76.985068   

           PROJECTNOTE                                        PROJECTNAME  \
0                  NaN                                             FedAid   
1                  NaN                                                NaN   
2                  NaN           Fedral Aid Citywide Pavement Restoration   
3  POKA-2003-B-0048-JJ  Citywide Pavement Restoration Contract (Local ...   
4                  NaN                                                NaN   

          PROJECTMGR  PROJECTID        TEAMREP                 EMAIL  \
0    Paul Stephenson       5834        O. Hill                   NaN   
1                NaN       2467            NaN                   NaN   
2  Mohamed Abdullahi       5486   Said Cherifi   said.cherifi@dc.gov   
3    Maduabuchi Udeh       5796  Michael Jelen  Michael.Jelen@dc.gov   
4                NaN       8237            NaN                   NaN   

   CONTACTPHONE  
0  202-671-4591  
1           NaN  
2  202-671-4611  
3  202-671-4542  
4           NaN  

[5 rows x 27 columns]

### 3.1.a  what is its shape and what does that mean?

In [32]:
data_frame.shape

(465, 27)

Shape will let us know number of columns and rows.As it has returned that there are 565 rows and 27 columns in the csv file above.

## Homework 3.2
 What are the number of rows in each 'QUADRANT' ?

In [110]:
q_unique=data_frame.QUADRANT.unique()
print q_unique
NW_count=data_frame.loc[data_frame['QUADRANT']=="NW"].shape
print NW_count
NE_count=data_frame.loc[data_frame['QUADRANT']=="NE"].shape
print NE_count
SW_count=data_frame.loc[data_frame['QUADRANT']=="SW"].shape
print SW_count
SE_count=data_frame.loc[data_frame['QUADRANT']=="SE"].shape
print SE_count
BN_count=data_frame.loc[data_frame['QUADRANT']=="BN"].shape
print BN_count

['NW' 'NE' 'SE' 'SW' 'BN']
(195, 27)
(163, 27)
(11, 27)
(84, 27)
(12, 27)


## Homework 3.3 - Array math demonstration 
For two arrarys  

a= [1,2,3,4]  type=float

b= [5,6,7,8]  type=float

Peform the following array operations using numpy   
( show both operational use of numpy and functional (example addition operation => + vs addition function => numbpy.add() )

### 3.3.1 addition a+b 

In [34]:
a=np.array([1,2,3,4],dtype=np.float)
b=np.array([5,6,7,8],dtype=np.float)
s=np.add(a, b)
print s
a+b

[  6.   8.  10.  12.]


array([  6.,   8.,  10.,  12.])

### 3.3.2 subtraction a-b

In [35]:
su=np.subtract(a,b)
print su
a-b

[-4. -4. -4. -4.]


array([-4., -4., -4., -4.])

### 3.3.3 multiplication a*b

In [36]:
m=np.multiply(a,b)
print m
a*b

[  5.  12.  21.  32.]


array([  5.,  12.,  21.,  32.])

### 3.3.4 divsion a/b

In [37]:
d=np.divide(a,b)
print d
a/b

[ 0.2         0.33333333  0.42857143  0.5       ]


array([ 0.2       ,  0.33333333,  0.42857143,  0.5       ])

### 3.3.5 modulo  a%b

In [38]:
mo=np.mod(a,b)
print mo
a%b

[ 1.  2.  3.  4.]


array([ 1.,  2.,  3.,  4.])

### 3.3.6 power  a^b

In [39]:
pow=np.float_power(a,b)
print pow
a**b

[  1.00000000e+00   6.40000000e+01   2.18700000e+03   6.55360000e+04]


array([  1.00000000e+00,   6.40000000e+01,   2.18700000e+03,
         6.55360000e+04])

## Homework 3.4
Find your own data and load it into a dataframe

In [140]:
import pandas as pd
df=pd.read_csv('C:\Users\AsusGo\Desktop\crime_2015.csv')
df.head(5)

#import numpy as np
#csv = np.genfromtxt('C:\Users\AsusGo\Desktop\crime_2015.csv', delimiter= ",")


X          Y       CCN                REPORT_DAT     SHIFT  METHOD  \
0 -76.979019  38.866897   6155556  2015-12-04T00:00:00.000Z  MIDNIGHT     GUN   
1 -77.022947  38.910389   9051410  2015-09-14T00:00:00.000Z  MIDNIGHT  OTHERS   
2 -76.980393  38.899616  11105327  2015-08-14T00:00:00.000Z  MIDNIGHT   KNIFE   
3 -76.933413  38.890401  11535619  2015-08-30T02:57:29.000Z  MIDNIGHT     GUN   
4 -77.072880  38.930449  14158527  2015-01-25T00:00:00.000Z  MIDNIGHT  OTHERS   

     OFFENSE                                     BLOCK    XBLOCK    YBLOCK  \
0   HOMICIDE       1900 - 1925 BLOCK OF 18TH STREET SE  401821.0  133230.0   
1   HOMICIDE        1500 - 1599 BLOCK OF 8TH STREET NW  398010.0  138058.0   
2  SEX ABUSE      1516 - 1699 BLOCK OF BENNING ROAD NE  401701.0  136862.0   
3    ROBBERY    4800 - 4899 BLOCK OF CENTRAL AVENUE NE  405777.0  135841.0   
4  SEX ABUSE  3000 - 3199 BLOCK OF WISCONSIN AVENUE NW  393681.0  140287.0   

     ...     CENSUS_TRACT VOTING_PRECINCT     XCOORD     YCOORD   LATITUDE  \
0    ...           7605.0    Precinct 112  401795.67  133225.61  38.866890   
1    ...           4901.0     Precinct 18  397984.28  137924.44  38.910382   
2    ...           7901.0     Precinct 81  401716.19  136822.69  38.899608   
3    ...           7804.0     Precinct 98  405802.40  135825.39  38.890393   
4    ...            400.0     Precinct 12  393647.71  140311.55  38.930441   

   LONGITUDE  BID                START_DATE                  END_DATE  \
0 -76.979017  NaN  2006-11-13T19:35:00.000Z  2006-11-13T19:38:00.000Z   
1 -77.022944  NaN  2009-04-18T01:05:00.000Z  2009-04-18T05:00:00.000Z   
2 -76.980391  NaN  2011-07-21T20:00:00.000Z  2011-07-21T23:00:00.000Z   
3 -76.933411  NaN  2015-08-30T02:14:49.000Z  2015-08-30T02:20:15.000Z   
4 -77.072878  NaN  2014-03-21T22:00:00.000Z  2014-03-21T22:05:00.000Z   

   OBJECTID  
0  71503077  
1  71529903  
2  71577169  
3  71588612  
4  71652614  

[5 rows x 26 columns]

## Homework 3.5
Provide an interesting analysis of the data columns ( frequency or averages )

Crimes done at different times

In [141]:
s_unique=df.SHIFT.unique()
print s_unique

['MIDNIGHT' 'DAY' 'EVENING']


In [142]:
MIDNIGHT_count=df.loc[df['SHIFT']=="MIDNIGHT"].shape
print MIDNIGHT_count
EVENING_count=df.loc[df['SHIFT']=="EVENING"].shape
print EVENING_count
DAY_count=df.loc[df['SHIFT']=="DAY"].shape
print DAY_count

(7023, 26)
(16253, 26)
(14042, 26)


According to the above results,it is determined that most of the crime took place in the evening shifts(16253).

In [143]:
o_unique=df.OFFENSE.unique()
print o_unique

['HOMICIDE' 'SEX ABUSE' 'ROBBERY' 'THEFT F/AUTO' 'BURGLARY' 'THEFT/OTHER'
 'MOTOR VEHICLE THEFT' 'ASSAULT W/DANGEROUS WEAPON' 'ARSON']


In [144]:
#import numpy as np
#print df.groupby(['SHIFT','OFFENSE','METHOD']).groups
#method_grouped=df.groupby('METHOD')
#print method_grouped['OFFENSE'].agg(np.mean)
import pandas as pd
import numpy as np

df = pd.DataFrame(np.random.rand(10,4),columns=['OFFENSE','SHIFT','METHOD']
df.plot.bar()

SyntaxError: invalid syntax (<ipython-input-144-35b2be05808d>, line 9)